### PR TITLE
Work around incompatibilities between V2 page header handling and zStandard compression in Parquet writer

### PR DESCRIPTION
**Pull request by etseidl**
## Description
In the version 2 Parquet page header, neither the repetition nor definition level data is compressed. The current Parquet writer achieves this by offsetting the input buffers passed to nvcomp to skip this level data. Doing so can lead to mis-aligned data being passed to nvcomp (for zstd, input currently must be aligned on a 4 byte boundary). This PR is a short-term fix that will print an error and exit if zStandard compression is used with V2 page headers. This also fixes an underestimation of the maximum V2 page header size.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
This PR is a temporary fix that should be in 24.02.  A proper fix will be ready for 24.04.

File: cpp/src/io/parquet/page_enc.cu [hunk 1]
```diff
@@ -57,6 +57,13 @@ constexpr int num_encode_warps  = encode_block_size / cudf::detail::warp_size;
 
 constexpr int rolling_idx(int pos) { return rolling_index<rle_buffer_size>(pos); }
 
+// max V1 header size (27 rounded up to 8 byte boundary)
+// also valid for dict page header (V1 or V2)
+constexpr int MAX_V1_HDR_SIZE = 32;
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:62]:
```suggestion
// max V1 header size 
// also valid for dict page header (V1 or V2)
constexpr int MAX_V1_HDR_SIZE = util::round_up_unsafe(27, 8);
```

File: cpp/src/io/parquet/page_enc.cu [hunk 2]
```diff
@@ -57,6 +57,13 @@ constexpr int num_encode_warps  = encode_block_size / cudf::detail::warp_size;
...
+
+// max V2 header size (49 rounded up to 8 byte boundary)
+constexpr int MAX_V2_HDR_SIZE = 56;
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:65]:
```suggestion
// max V2 header size
constexpr int MAX_V2_HDR_SIZE = util::round_up_unsafe(49, 8);
```

File: cpp/src/io/parquet/writer_impl.cu [hunk 3]
```diff
@@ -2184,6 +2184,9 @@ writer::impl::impl(std::vector<std::unique_ptr<data_sink>> sinks,
   if (options.get_metadata()) {
     _table_meta = std::make_unique<table_input_metadata>(*options.get_metadata());
   }
+  if (_write_v2_headers and _compression == Compression::ZSTD) {
+    CUDF_FAIL("V2 page headers cannot be used with ZSTD compression");
```
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:2188]: we need the same check for the chunked case as well

File: cpp/src/io/parquet/writer_impl.cu
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:1]: can you please add tests? something like
```
TEST_F(ParquetWriterStressTest, ZstdWithV2Header)
{
  auto expected = table_view{};

  cudf::io::parquet_writer_options out_opts =
    cudf::io::parquet_writer_options::builder(cudf::io::sink_info{"unused.pq"}, expected)
      .compression(cudf::io::compression_type::ZSTD)
      .write_v2_headers(true);
  EXPECT_THROW(cudf::io::write_parquet(out_opts), cudf::logic_error);
}
```
should do it.

**Review by vuule — CHANGES_REQUESTED** [review]
Thank you for the workaround!

File: cpp/src/io/parquet/writer_impl.cu (hunk 3 above)
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl.cu:2188]: Oops. Or should the test (and the lines above it) go in `init_state()` instead? (And maybe change the name of `init_state`).

Comment (vuule) [reply cpp/src/io/parquet/writer_impl.cu]: Looks like I'm late with the answer:)
Side note: the current code in `init_state` should not be there (breaks the strong exception guarantee)

File: cpp/src/io/parquet/page_enc.cu [hunk 4]
```diff
@@ -534,6 +541,9 @@ CUDF_KERNEL void __launch_bounds__(128)
   uint32_t const t          = threadIdx.x;
   auto const data_page_type = write_v2_headers ? PageType::DATA_PAGE_V2 : PageType::DATA_PAGE;
 
+  // Max page header size excluding statistics
+  auto const max_data_page_hdr_size = write_v2_headers ? MAX_V2_HDR_SIZE : MAX_V1_HDR_SIZE;
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:545]: this is unrelated to the WAR, right?

**Review by vuule — APPROVED** [review]

File: cpp/src/io/parquet/page_enc.cu (hunk 4 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu:545]: Yes, it's replacing (incorrect as it turns out) magic numbers I noticed while tracking down the bug.

File: cpp/src/io/parquet/writer_impl.cu [hunk 5]
```diff
@@ -2220,6 +2220,10 @@ writer::impl::~impl() { close(); }
 
 void writer::impl::init_state()
 {
+  if (_write_v2_headers and _compression == Compression::ZSTD) {
+    CUDF_FAIL("V2 page headers cannot be used with ZSTD compression");
+  }
```
Comment (vuule) [inline cpp/src/io/parquet/writer_impl.cu:2225]: Very optional suggestion :D 
```suggestion
    CUDF_EXPECTS(not _write_v2_headers or _compression != Compression::ZSTD", V2 page headers cannot be used with ZSTD compression");
```

**Review by davidwendt — APPROVED** [review]
